consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+
+    fn get_block_proof(&self, _block: &Block) -> u128 {
+        //TODO: We have to make correct one
+        1
+    }
 
-// impl<S: BlockchainStorage> Consensus<S> {
-//     pub fn process_block(block: Block) {
+    fn add_to_block_index(
+        &mut self,
+        block: &Block,
+        best_block_id: &Option<Id<Block>>,
+    ) -> Result<BlockIndex, BlockError> {
+        let prev_block_index = if best_block_id.is_none() && block.is_genesis(self.chain_config) {
+            // Genesis case
+            None
+        } else {
+            block.prev_block_id().map_or(Err(BlockError::Orphan), |prev_block| {
+                self.db_tx.get_block_index(&prev_block).map_err(BlockError::from)
+            })?
+        };
+        // Set the block height
+        let height = prev_block_index.as_ref().map_or(BlockHeight::zero(), |prev_block_index| {
+            prev_block_index.get_block_height().next_height()
+        });
 
-//     }
-// }
+        // Set Time Max
+        let time_max = prev_block_index.as_ref().map_or(block.block_time(), |prev_block_index| {
+            std::cmp::max(prev_block_index.get_block_time_max(), block.block_time())
+        });
+
+        // Set Chain Trust
+        let chain_trust = prev_block_index
+            .map_or(0, |prev_block_index| prev_block_index.get_chain_trust())
+            + self.get_block_proof(block);
+        let block_index = BlockIndex::new(block, chain_trust, height, time_max);
+        Ok(block_index)
+    }
+
+    #[allow(dead_code)]
+    fn accept_block(
+        &mut self,
+        block: &Block,
+        best_block_id: &Option<Id<Block>>,
+    ) -> Result<BlockIndex, BlockError> {
+        let block_index = self.add_to_block_index(block, best_block_id)?;
+        self.check_block_index(&block_index)?;
+        self.db_tx.set_block_index(&block_index).map_err(BlockError::from)?;
+        self.db_tx.add_block(block).map_err(BlockError::from)?;
+        Ok(block_index)
+    }
+
+    fn block_exists(&self, id: &Id<Block>) -> Result<bool, BlockError> {
+        Ok(self.db_tx.get_block_index(id)?.is_some())
+    }
+
+    fn check_block_index(&self, block_index: &BlockIndex) -> Result<(), BlockError> {
+        // BlockIndex is already known or block exists
+        if self.db_tx.get_block_index(block_index.get_block_id())?.is_some()
+            || self.block_exists(block_index.get_block_id())?
+        {
+            return Err(BlockError::Unknown);
+        }
+        // TODO: Will be expanded
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_block_detail(
+        &self,
+        block: &Block,
+        block_source: BlockSource,
+    ) -> Result<(), BlockError> {
+        // MerkleTree root
+        let merkle_tree_root = block.merkle_root();
+        calculate_tx_merkle_root(block.transactions()).map_or(
+            Err(BlockError::Unknown),
+            |merkle_tree| {
+                if merkle_tree_root != merkle_tree {
+                    Err(BlockError::Unknown)
```

Comment:
I think `BlockError::Unknown` is overused and there should be more errors

Reply:
The idea was that we use `Unknown` for now as a main error and when the priority tasks will be done, we decide about which errors should use through the code. 

Also, we have the issue for it: https://github.com/mintlayer/mintlayer-core/issues/126